File: q0541/q0541.cpp
#include <string>
using namespace std;

class Solution
{
public:
    string reverseStr(string s, int k)
    {
        int n = s.size();
        for (int l = 0; l < n; l += 2 * k)
        {
            int r = min(l + k, n);
            // Or replace the following lines with:
            // reverse(s.begin()+l, s.begin()+r);
            for (int i = l, j = r - 1; i < j; i++, j--)
            {
                char tmp = s[i];
                s[i] = s[j];
                s[j] = tmp;
            }
        }
        return s;
    }
};

File: q0541/q0541.java
class Solution {
    public String reverseStr(String s, int k) {
        char[] charArr = s.toCharArray();

        // Find the left bound of the substring to be reversed
        for (int l = 0; l < s.length(); l += 2 * k) {

            // Find the right bound of the substring to be reversed
            int r = Math.min(s.length() - 1, l + k - 1);

            // Reverse the substring bounded by [i, j) in place
            for (int i = l, j = r; i < j; i++, j--) {
                char temp = charArr[i];
                charArr[i] = charArr[j];
                charArr[j] = temp;
            }
        }
        return new String(charArr);
    }
}